src/metabase/db/liquibase.clj

```diff
@@ -127,16 +128,20 @@
   [conn-or-data-source :- [:or (ms/InstanceOfClass java.sql.Connection) (ms/InstanceOfClass javax.sql.DataSource)]
    f                   :- fn?]
   ;; Custom migrations use toucan2, so we need to make sure it uses the same connection with liquibase
-  (binding [t2.conn/*current-connectable* conn-or-data-source]
-    (if (instance? java.sql.Connection conn-or-data-source)
-      (f (->> conn-or-data-source liquibase-connection database (liquibase conn-or-data-source)))
-      ;; closing the `LiquibaseConnection`/`Database` closes the parent JDBC `Connection`, so only use it in combination
-      ;; with `with-open` *if* we are opening a new JDBC `Connection` from a JDBC spec. If we're passed in a `Connection`,
-      ;; it's safe to assume the caller is managing its lifecycle.
-      (with-open [conn           (.getConnection ^javax.sql.DataSource conn-or-data-source)
-                  liquibase-conn (liquibase-connection conn)
-                  database       (database liquibase-conn)]
-        (f (liquibase conn database))))))
+  (let [f* (fn [liquibase]
+             ;; trigger liquibase to create databasechangelog tables if needed
+             (.checkLiquibaseTables liquibase false (.getDatabaseChangeLog liquibase) nil nil)
```

Comment:
this is a work around for this bug https://github.com/liquibase/liquibase/issues/5537

Reply:
> without updating any checksums

wdym by this?

Reply:
> I think it's even worth extracting to ensure-liquibase-tables-exist function to encapsulate the misleadingly named checkLiquibaseTables method and its noisy unused arguments.

I'm in favor of extracting functions for reusing purposes; if for readability, I think the comment already does its justice here.

I also think we should keep the right number of "helpers" functions, cause if someone come in and read this for the first time, it's easy to get loss between helpers function vs main function that capture the business logic.

Reply:
> Without updating any checksums

I'm referencing to the `updateExistingNullChecksums` parameter that you set to false, and which makes the last two arguments irrelevant and thus ni.

> I'm in favor of extracting functions for reusing purposes; if for readability, the comment already does its justice here.

In my opinion comments should be used sparingly to explain *why* things are done, and only seldomly for "what" and even less the "how". The latter kinds of comments are [typically](https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/) code smells suggesting code could be more readable. 

In this instance I think the "what" comment is a crutch, and the better approach is to introduce a good name in the form of a helper function.

> I also think we should keep the right number of "helpers" functions.

From a bunch of similar discussions with you and Cal, I'm starting to suspect we have different opinions on "the right number" and code readability. Here's a wall of text about my take on code structure. 

This is just meant food for thought, ultimately a consistent and cohesive approach trumps a holy crusade for the one true style  ⛪ 

1. A key claim of FP is that breaking complex things down to the composition of "simple parts" makes reasoning easier, and yields code that is more robust to changing requirements.
2. [Simple things don't mix concerns, they do "one thing", and have "on reason to change".](https://blog.cleancoder.com/uncle-bob/2014/05/08/SingleReponsibilityPrinciple.html)
3. [Simple things don't mix levels of abstraction](http://principles-wiki.net/principles:single_level_of_abstraction). Jumping between abstraction levels is mentally taxing.
4. Removing duplication is [not a goal in itself](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction), it just  

> If someone comes in and reads this for the first time, it's easy to get lost between the helpers function ...

I think this is a culturally relative opinion. To be a bit snarky: I'm the new guy here, and giving you "more objective" feedback on what would have make the code easier to follow. 😄 

To maybe add some heft to my cheekiness - I was the custodian for a very large, old and non-trivial codebase over a long period of time, with multiple large teams working on it, and plenty of engineer turnover. My opinions reflect what seemed to work in my experience to tame this complexity and made code more approachable and flexible to change. YMMV.

Personally, I find it much easier to jump between simple functions than to comprehend large procedures. 

At the end of the day Clojure runs *a lot* of layers down until your reach the turtles, and all this code in particular is calling down to a pretty convoluted Java package. Clarity comes from having the right abstractions, not from to surfacing every detail. Ultimately I want to minimize the time taken to drill down to the code of interest, and minimize the amount of code read on the way there, rather then reading things from the bottom up in big procedures.

> vs the main function that captures the business logic.

I think part of the problems is that we're often trying to pack the business logic into these "main" functions, versus a "function core" approach where logic is kept separate to imperative, orchestrating code.

Reply:
There is a lot to chew on here, thank you for writing down your reasoning. I'll add those resources to my to-read list for my weekend mornings :).

for now, I'll update the comment as you suggested and leave the call as is.